In [2]:
import numpy as np
import pandas as pd
from scipy.stats import norm

# Задание 4.14

Нулевая гипотеза $H_0 : \sigma = \sigma_0$
$H_1 : \sigma > \sigma_0$

In [6]:
data = np.array([39.4, 39.6, 39.6, 39.2, 39.4, 39.3, 39.1, 39.6, 39.5, 39.6, 39.4,\
                 39.6, 39.4, 39.4, 39.5, 39.5, 39.4, 39.6, 39.5, 39.4, 39.1, 39.1, 39.7, 39.3, 39.1, 39])
sigma_0 = 0.15
mu = 39.4
print(data.size)

26


Вычислим $(n-1)s^2$

In [4]:
first = (data - mu).sum()
s_squared = ((data - mu) ** 2).sum() - first ** 2 / data.size
s_squared

0.9496153846153864

Квантиль распределения $\chi^2$ с n - 1 = 25 степенями свободы $\chi_{0.95}^2(25)=37.65$

In [8]:
Z = s_squared / sigma_0 ** 2
Z

42.20512820512829

Так как 42.205 > 37.65 то основная гипотеза $ H_0 $ отклоняется и принимается гипотеза $ H_1$

# Задание 5.14
Проверить гипотезу о нормальном распределении
$\alpha = 0.01$

In [3]:
data = pd.DataFrame({"from": [42.8, 43, 43.2, 43.4, 43.6, 43.8, 44, 44.2, 44.4, 44.6],\
                     "to": [43, 43.2, 43.4, 43.6, 43.8, 44, 44.2, 44.4, 44.6, 44.8],\
                     "n": [5, 16, 84, 113, 119, 78, 50, 19, 11, 5]})
C = (42.8 + 44.8) / 2
data['avg'] = (data['from'] + data['to']) / 2

In [4]:
mu = ((data['avg'] - C) * data['n']).sum() / data['n'].sum() + C
sigma_squared = np.sqrt(((data['avg'] - C) ** 2 * data['n']).sum() / data['n'].sum() - (mu - C) ** 2)
vals = []
for index, row in data.iterrows():
    vals.append(norm.cdf((row['to'] - mu)/sigma_squared) - norm.cdf((row['from'] - mu)/sigma_squared))
data['p_i'] = vals
data['u_i'] = (data['n'] - data['p_i'] * data['n'].sum()) ** 2 / (data['p_i'] * data['n'].sum())
print("U =",data['u_i'].sum())

U = 31.526799013487146


Квантиль распределения $\chi^2$ с r = k - l - 1 = 10 - 2 - 1 = 7 степенями свободы $\chi_{0.99}^2(25)=18,47531$, U > $\chi^2$, то есть гипотеза о нормальном распределении отклоняется